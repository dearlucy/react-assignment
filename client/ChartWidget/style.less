@import '../styles/variables';

.container {
  background-color: @widget-background-color;
  color: @widget-color;
  padding: @padding;
  order: 2;
}

.title {
  text-align: center;
}

.btn {
  padding: 4px;
  font-size: .9rem;
  border: 1px solid #61575d;
  background-color: #432e3b;
  color: #c4bec2;

  &:focus {
    outline: none;
    /* in real project would be better to add a script, to support a11y */
  }
}