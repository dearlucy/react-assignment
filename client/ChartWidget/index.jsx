import React, { Component } from "react";
import {
  LineChart,
  Line,
  CartesianGrid,
  XAxis,
  YAxis,
  Tooltip,
} from "recharts";
import moment from "moment";
import style from "./style.less";

class Chart extends Component {
  constructor() {
    super();
    this.state = {
      chartData: [],
      chartDataThisYear: [],
      showCumulitive: false,
    };

    this.handleToggle = this.handleToggle.bind(this);
    this.countCumulitive = this.countCumulitive.bind(this);
  }

  componentDidMount() {
    const url = "http://localhost:8000/series";
    fetch(url)
      .then((response) => response.json())
      .then((data) =>
        this.setState((state) => {
          return { state, chartData: data };
        })
      )
      .catch((error) => {
        console.error('Error:', error);
      });
  }

  countCumulitive() {
    let thisYear = moment().format("YYYY");
    let thisYearData = this.state.chartData.filter((item) =>
      item.date.includes(thisYear)
    );
    let cumulitive = thisYearData
      .map((obj) => obj.value)
      .reduce((acc, cur) => acc + cur);
    thisYearData.push({
      date: thisYear,
      value: cumulitive,
    });

    this.setState((state) => {
      return {
        state,
        chartDataThisYear: thisYearData,
      };
    });
  }

  handleToggle() {
    this.setState((prevState) => {
      return {
        prevState,
        showCumulitive: !prevState.showCumulitive,
      };
    });

    if (!this.state.chartDataThisYear.length) {
      this.countCumulitive();
    }
  }

  render() {
    return (
      <div className={style.container}>
        <button className={style.btn} onClick={this.handleToggle}>
          {this.state.showCumulitive ? 'Show values per month' : 'Show cumulitive'}
        </button>
        <h3 className={style.title}>
          {this.state.showCumulitive
            ? "Cumulative value for this year"
            : "Values per month 2019 - 2020"
            }
        </h3>
          <LineChart
          width={800}
          height={400}
          data={
            this.state.showCumulitive
              ? this.state.chartDataThisYear
              : this.state.chartData
          }
          margin={{ top: 5, right: 20, bottom: 5, left: 0 }}
        >
          <Line type="linear" dataKey="value" stroke="#8884d8" />
          <CartesianGrid stroke="#ccc" strokeDasharray="5 5" />
          <XAxis dataKey="date" />
          <YAxis />
          <Tooltip />
        </LineChart>
      </div>
    );
  }
}

export default Chart;
