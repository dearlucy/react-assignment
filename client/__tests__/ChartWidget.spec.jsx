import React from 'react';
import { shallow } from 'enzyme';
import ChartWidget from '../ChartWidget';
import {
  LineChart,
  Line,
  CartesianGrid,
  XAxis,
  YAxis,
  Tooltip,
} from "recharts";


describe("ChartWidget component", () => {
  let wrapper = shallow(<ChartWidget />);
  it("Should render component", () => {
    expect(wrapper).toMatchSnapshot();
  });
});

describe("ChartWidget renders components", () => {
  let wrapper;

  beforeEach(() => (wrapper = shallow(<ChartWidget />)));

  it("renders title", () => {
    expect(wrapper.find("h3").length).toBe(1);
  });

  it("renders button", () => {
    expect(wrapper.find("button").length).toBe(1);
  });

  it("Renders LineChart", () => {
    expect(wrapper.find(LineChart).length).toBe(1);
  });

  it("Renders Line", () => {
    expect(wrapper.find(Line).length).toBe(1);
  });

  it("Renders CartesianGrid", () => {
    expect(wrapper.find(CartesianGrid).length).toBe(1);
  });

  it("Renders XAxis", () => {
    expect(wrapper.find(XAxis).length).toBe(1);
  });

  it("Renders YAxis", () => {
    expect(wrapper.find(YAxis).length).toBe(1);
  });

  it("Renders Tooltip", () => {
    expect(wrapper.find(Tooltip).length).toBe(1);
  });

});


const componentDidMountSpy = jest.spyOn(ChartWidget.prototype, 'componentDidMount');

describe("Lifecycle method", () => {
  let wrapper = shallow(<ChartWidget />, { lifecycleExperimental: true });
  it("componentDidMoun called only once", () => {
      expect(componentDidMountSpy).toHaveBeenCalledTimes(1);
  })

  it("componentDidMoun not called twice", () => {
    expect(componentDidMountSpy).not.toHaveBeenCalledTimes(2);
  })
})