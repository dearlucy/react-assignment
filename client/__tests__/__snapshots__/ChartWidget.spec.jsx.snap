// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartWidget component Should render component 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    className={undefined}
>
    <button
        className={undefined}
        onClick={[Function]}
    >
        Show cumulitive
    </button>
    <h3
        className={undefined}
    >
        Values per month 2019 - 2020
    </h3>
    <LineChart
        barCategoryGap="10%"
        barGap={4}
        data={Array []}
        height={400}
        layout="horizontal"
        margin={
            Object {
                "bottom": 5,
                "left": 0,
                "right": 20,
                "top": 5,
              }
        }
        reverseStackOrder={false}
        stackOffset="none"
        width={800}
    >
        <Line
            activeDot={true}
            animateNewValues={true}
            animationBegin={0}
            animationDuration={1500}
            animationEasing="ease"
            connectNulls={false}
            dataKey="value"
            dot={true}
            fill="#fff"
            hide={false}
            isAnimationActive={true}
            legendType="line"
            onAnimationEnd={[Function]}
            onAnimationStart={[Function]}
            points={Array []}
            stroke="#8884d8"
            strokeWidth={1}
            type="monotone"
            xAxisId={0}
            yAxisId={0}
        />
        <CartesianGrid
            fill="none"
            horizontal={true}
            horizontalFill={Array []}
            horizontalPoints={Array []}
            stroke="#ccc"
            strokeDasharray="5 5"
            vertical={true}
            verticalFill={Array []}
            verticalPoints={Array []}
        />
        <XAxis
            allowDataOverflow={false}
            allowDecimals={true}
            allowDuplicatedCategory={true}
            dataKey="date"
            domain={
                Array [
                    0,
                    "auto",
                  ]
            }
            height={30}
            hide={false}
            mirror={false}
            orientation="bottom"
            padding={
                Object {
                    "left": 0,
                    "right": 0,
                  }
            }
            reversed={false}
            scale="auto"
            tickCount={5}
            type="category"
            width={0}
            xAxisId={0}
        />
        <YAxis
            allowDataOverflow={false}
            allowDecimals={true}
            allowDuplicatedCategory={true}
            domain={
                Array [
                    0,
                    "auto",
                  ]
            }
            height={0}
            hide={false}
            mirror={false}
            orientation="left"
            padding={
                Object {
                    "bottom": 0,
                    "top": 0,
                  }
            }
            reversed={false}
            scale="auto"
            tickCount={5}
            type="number"
            width={60}
            yAxisId={0}
        />
        <Tooltip
            active={false}
            allowEscapeViewBox={
                Object {
                    "x": false,
                    "y": false,
                  }
            }
            animationDuration={400}
            animationEasing="ease"
            contentStyle={Object {}}
            coordinate={
                Object {
                    "x": 0,
                    "y": 0,
                  }
            }
            cursor={true}
            cursorStyle={Object {}}
            filterNull={true}
            isAnimationActive={true}
            itemStyle={Object {}}
            labelStyle={Object {}}
            offset={10}
            separator=" : "
            useTranslate3d={false}
            viewBox={
                Object {
                    "x1": 0,
                    "x2": 0,
                    "y1": 0,
                    "y2": 0,
                  }
            }
            wrapperStyle={Object {}}
        />
    </LineChart>
</div>,
  "nodes": Array [
    <div
      className={undefined}
>
      <button
            className={undefined}
            onClick={[Function]}
      >
            Show cumulitive
      </button>
      <h3
            className={undefined}
      >
            Values per month 2019 - 2020
      </h3>
      <LineChart
            barCategoryGap="10%"
            barGap={4}
            data={Array []}
            height={400}
            layout="horizontal"
            margin={
                  Object {
                        "bottom": 5,
                        "left": 0,
                        "right": 20,
                        "top": 5,
                      }
            }
            reverseStackOrder={false}
            stackOffset="none"
            width={800}
      >
            <Line
                  activeDot={true}
                  animateNewValues={true}
                  animationBegin={0}
                  animationDuration={1500}
                  animationEasing="ease"
                  connectNulls={false}
                  dataKey="value"
                  dot={true}
                  fill="#fff"
                  hide={false}
                  isAnimationActive={true}
                  legendType="line"
                  onAnimationEnd={[Function]}
                  onAnimationStart={[Function]}
                  points={Array []}
                  stroke="#8884d8"
                  strokeWidth={1}
                  type="monotone"
                  xAxisId={0}
                  yAxisId={0}
            />
            <CartesianGrid
                  fill="none"
                  horizontal={true}
                  horizontalFill={Array []}
                  horizontalPoints={Array []}
                  stroke="#ccc"
                  strokeDasharray="5 5"
                  vertical={true}
                  verticalFill={Array []}
                  verticalPoints={Array []}
            />
            <XAxis
                  allowDataOverflow={false}
                  allowDecimals={true}
                  allowDuplicatedCategory={true}
                  dataKey="date"
                  domain={
                        Array [
                              0,
                              "auto",
                            ]
                  }
                  height={30}
                  hide={false}
                  mirror={false}
                  orientation="bottom"
                  padding={
                        Object {
                              "left": 0,
                              "right": 0,
                            }
                  }
                  reversed={false}
                  scale="auto"
                  tickCount={5}
                  type="category"
                  width={0}
                  xAxisId={0}
            />
            <YAxis
                  allowDataOverflow={false}
                  allowDecimals={true}
                  allowDuplicatedCategory={true}
                  domain={
                        Array [
                              0,
                              "auto",
                            ]
                  }
                  height={0}
                  hide={false}
                  mirror={false}
                  orientation="left"
                  padding={
                        Object {
                              "bottom": 0,
                              "top": 0,
                            }
                  }
                  reversed={false}
                  scale="auto"
                  tickCount={5}
                  type="number"
                  width={60}
                  yAxisId={0}
            />
            <Tooltip
                  active={false}
                  allowEscapeViewBox={
                        Object {
                              "x": false,
                              "y": false,
                            }
                  }
                  animationDuration={400}
                  animationEasing="ease"
                  contentStyle={Object {}}
                  coordinate={
                        Object {
                              "x": 0,
                              "y": 0,
                            }
                  }
                  cursor={true}
                  cursorStyle={Object {}}
                  filterNull={true}
                  isAnimationActive={true}
                  itemStyle={Object {}}
                  labelStyle={Object {}}
                  offset={10}
                  separator=" : "
                  useTranslate3d={false}
                  viewBox={
                        Object {
                              "x1": 0,
                              "x2": 0,
                              "y1": 0,
                              "y2": 0,
                            }
                  }
                  wrapperStyle={Object {}}
            />
      </LineChart>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Chart />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Chart {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "countCumulitive": [Function],
        "handleToggle": [Function],
        "props": Object {},
        "refs": Object {},
        "state": Object {
          "chartData": Array [],
          "chartDataThisYear": Array [],
          "showCumulitive": false,
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          className={undefined}
>
          <button
                    className={undefined}
                    onClick={[Function]}
          >
                    Show cumulitive
          </button>
          <h3
                    className={undefined}
          >
                    Values per month 2019 - 2020
          </h3>
          <LineChart
                    barCategoryGap="10%"
                    barGap={4}
                    data={Array []}
                    height={400}
                    layout="horizontal"
                    margin={
                              Object {
                                        "bottom": 5,
                                        "left": 0,
                                        "right": 20,
                                        "top": 5,
                                      }
                    }
                    reverseStackOrder={false}
                    stackOffset="none"
                    width={800}
          >
                    <Line
                              activeDot={true}
                              animateNewValues={true}
                              animationBegin={0}
                              animationDuration={1500}
                              animationEasing="ease"
                              connectNulls={false}
                              dataKey="value"
                              dot={true}
                              fill="#fff"
                              hide={false}
                              isAnimationActive={true}
                              legendType="line"
                              onAnimationEnd={[Function]}
                              onAnimationStart={[Function]}
                              points={Array []}
                              stroke="#8884d8"
                              strokeWidth={1}
                              type="monotone"
                              xAxisId={0}
                              yAxisId={0}
                    />
                    <CartesianGrid
                              fill="none"
                              horizontal={true}
                              horizontalFill={Array []}
                              horizontalPoints={Array []}
                              stroke="#ccc"
                              strokeDasharray="5 5"
                              vertical={true}
                              verticalFill={Array []}
                              verticalPoints={Array []}
                    />
                    <XAxis
                              allowDataOverflow={false}
                              allowDecimals={true}
                              allowDuplicatedCategory={true}
                              dataKey="date"
                              domain={
                                        Array [
                                                  0,
                                                  "auto",
                                                ]
                              }
                              height={30}
                              hide={false}
                              mirror={false}
                              orientation="bottom"
                              padding={
                                        Object {
                                                  "left": 0,
                                                  "right": 0,
                                                }
                              }
                              reversed={false}
                              scale="auto"
                              tickCount={5}
                              type="category"
                              width={0}
                              xAxisId={0}
                    />
                    <YAxis
                              allowDataOverflow={false}
                              allowDecimals={true}
                              allowDuplicatedCategory={true}
                              domain={
                                        Array [
                                                  0,
                                                  "auto",
                                                ]
                              }
                              height={0}
                              hide={false}
                              mirror={false}
                              orientation="left"
                              padding={
                                        Object {
                                                  "bottom": 0,
                                                  "top": 0,
                                                }
                              }
                              reversed={false}
                              scale="auto"
                              tickCount={5}
                              type="number"
                              width={60}
                              yAxisId={0}
                    />
                    <Tooltip
                              active={false}
                              allowEscapeViewBox={
                                        Object {
                                                  "x": false,
                                                  "y": false,
                                                }
                              }
                              animationDuration={400}
                              animationEasing="ease"
                              contentStyle={Object {}}
                              coordinate={
                                        Object {
                                                  "x": 0,
                                                  "y": 0,
                                                }
                              }
                              cursor={true}
                              cursorStyle={Object {}}
                              filterNull={true}
                              isAnimationActive={true}
                              itemStyle={Object {}}
                              labelStyle={Object {}}
                              offset={10}
                              separator=" : "
                              useTranslate3d={false}
                              viewBox={
                                        Object {
                                                  "x1": 0,
                                                  "x2": 0,
                                                  "y1": 0,
                                                  "y2": 0,
                                                }
                              }
                              wrapperStyle={Object {}}
                    />
          </LineChart>
</div>,
        "_debugID": 2,
        "_renderedOutput": <div
          className={undefined}
>
          <button
                    className={undefined}
                    onClick={[Function]}
          >
                    Show cumulitive
          </button>
          <h3
                    className={undefined}
          >
                    Values per month 2019 - 2020
          </h3>
          <LineChart
                    barCategoryGap="10%"
                    barGap={4}
                    data={Array []}
                    height={400}
                    layout="horizontal"
                    margin={
                              Object {
                                        "bottom": 5,
                                        "left": 0,
                                        "right": 20,
                                        "top": 5,
                                      }
                    }
                    reverseStackOrder={false}
                    stackOffset="none"
                    width={800}
          >
                    <Line
                              activeDot={true}
                              animateNewValues={true}
                              animationBegin={0}
                              animationDuration={1500}
                              animationEasing="ease"
                              connectNulls={false}
                              dataKey="value"
                              dot={true}
                              fill="#fff"
                              hide={false}
                              isAnimationActive={true}
                              legendType="line"
                              onAnimationEnd={[Function]}
                              onAnimationStart={[Function]}
                              points={Array []}
                              stroke="#8884d8"
                              strokeWidth={1}
                              type="monotone"
                              xAxisId={0}
                              yAxisId={0}
                    />
                    <CartesianGrid
                              fill="none"
                              horizontal={true}
                              horizontalFill={Array []}
                              horizontalPoints={Array []}
                              stroke="#ccc"
                              strokeDasharray="5 5"
                              vertical={true}
                              verticalFill={Array []}
                              verticalPoints={Array []}
                    />
                    <XAxis
                              allowDataOverflow={false}
                              allowDecimals={true}
                              allowDuplicatedCategory={true}
                              dataKey="date"
                              domain={
                                        Array [
                                                  0,
                                                  "auto",
                                                ]
                              }
                              height={30}
                              hide={false}
                              mirror={false}
                              orientation="bottom"
                              padding={
                                        Object {
                                                  "left": 0,
                                                  "right": 0,
                                                }
                              }
                              reversed={false}
                              scale="auto"
                              tickCount={5}
                              type="category"
                              width={0}
                              xAxisId={0}
                    />
                    <YAxis
                              allowDataOverflow={false}
                              allowDecimals={true}
                              allowDuplicatedCategory={true}
                              domain={
                                        Array [
                                                  0,
                                                  "auto",
                                                ]
                              }
                              height={0}
                              hide={false}
                              mirror={false}
                              orientation="left"
                              padding={
                                        Object {
                                                  "bottom": 0,
                                                  "top": 0,
                                                }
                              }
                              reversed={false}
                              scale="auto"
                              tickCount={5}
                              type="number"
                              width={60}
                              yAxisId={0}
                    />
                    <Tooltip
                              active={false}
                              allowEscapeViewBox={
                                        Object {
                                                  "x": false,
                                                  "y": false,
                                                }
                              }
                              animationDuration={400}
                              animationEasing="ease"
                              contentStyle={Object {}}
                              coordinate={
                                        Object {
                                                  "x": 0,
                                                  "y": 0,
                                                }
                              }
                              cursor={true}
                              cursorStyle={Object {}}
                              filterNull={true}
                              isAnimationActive={true}
                              itemStyle={Object {}}
                              labelStyle={Object {}}
                              offset={10}
                              separator=" : "
                              useTranslate3d={false}
                              viewBox={
                                        Object {
                                                  "x1": 0,
                                                  "x2": 0,
                                                  "y1": 0,
                                                  "y2": 0,
                                                }
                              }
                              wrapperStyle={Object {}}
                    />
          </LineChart>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Chart />,
}
`;
